### PR TITLE
Adds Canonical, Url, Uri search support

### PR DESCRIPTION
**Pull request by brendankowitz**
## Description

- Canonical => Uri Support
- Url => Uri support
- Uri/Url => Reference support (note: not canonical)
- Adds ability for tests to detect if a Search Param is supported in CapabilityStatement

### Note on `_profile` in R4:
In reference to the [spec](https://www.hl7.org/fhir/search.html#uri) where `_profile` is a canonical converting to a `Uri` search type, in this case we will remove the canonical components (if present) and index the Uri only. When searching with _profile the same behavior applies where the canonical components will not interact with the search. We have raised ticket [FHIR-30514](https://jira.hl7.org/browse/FHIR-30514) to clarify this.

Reasoning for ignoring the canonical components when converting from Canonical => Uri revolves around the terminology:

>  SHOULD support automatically detecting a |[version] portion as part of the search parameter, and interpreting that portion as a search on the version

In future work (#1566), using a canonical with a `reference` search type, where the [target entity](https://www.hl7.org/fhir/references.html#canonical) is one that exists as part of the Definition pattern, the `|4` component will automatically be converted to the search term `&version=4` [as designed](https://build.fhir.org/resource.html#versions).

## Related issues
Refs #882

## Testing
Updates unit and integration tests

Adds the following searches (R4):
- /Observation?_profile=https://uri
- /ConceptMap?source=ValueSet/1 (note: Canonical => Reference is not implemented yet)
- /ConceptMap?target=ValueSet/1 (note: Canonical => Reference is not implemented yet)
- /DocumentReference?location=http://attatchment-uri
- /MessageHeader?destination=http://uri
- /MessageHeader?source=http://uri
- /PlanDefinition?definition=PlanDefinition/1 (note: Canonical => Reference is not implemented yet)
- /Subscription?url=http://channel-uri

## FHIR Team Checklist
- [x] **Update the title** of the PR to be succinct and less than 50 characters
- [x] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [x] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [x] Tag the PR with **Azure API for FHIR** if this will release to the managed service
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch

### CODE REVIEW

**Comment by brendankowitz** [comment]
@feordin, @johnstairs thanks for reviewing this work previously. I've updated the approach in the PR description, let me know if this makes sense or you have questions?

//cc @CaitlinV39

File: src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/UriSearchValue.cs [hunk 1]
```diff
@@ -38,43 +88,58 @@ public UriSearchValue(string uri)
...
-
             var uriSearchValueOther = other as UriSearchValue;
 
             if (uriSearchValueOther == null)
             {
                 return false;
             }
 
-            return Uri == uriSearchValueOther.Uri;
+            return ToString() == uriSearchValueOther.ToString();
```
Comment (feordin) [inline src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/UriSearchValue.cs:124]: Are the canonical components case sensitive for the comparison?  Should this be a case insensitive comparison?

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/CanonicalSearchTests.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+using Microsoft.Health.Fhir.Core.Models;
+using Microsoft.Health.Fhir.Tests.Common.Extensions;
+using Microsoft.Health.Fhir.Tests.Common.FixtureParameters;
+using Xunit;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.Health.Fhir.Tests.E2E.Rest.Search
+{
+    [HttpIntegrationFixtureArgumentSets(DataStore.All, Format.Json)]
+    public class CanonicalSearchTests : SearchTestsBase<CanonicalSearchTestFixture>
```
Comment (feordin) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/CanonicalSearchTests.cs:19]: I was looking for a test case with a canonical style URL that was being treated as a reference and then used in a search.  Perhaps we have it, and I just can't find it?

Comment (brendankowitz) [reply test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/CanonicalSearchTests.cs:19]: This PR doesn't cover Canonical => Reference, I plan to address this separately in #1566.

File: src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/UriSearchValue.cs (hunk 1 above)
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/UriSearchValue.cs:124]: It does say: `URLs are always considered to be case-sensitive` (https://www.hl7.org/fhir/references.html#literal)
However, it probably helps to be explicit about this.
```suggestion
            return string.Equals(ToString(), uriSearchValueOther.ToString(), StringComparison.Ordinal);
```

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/CanonicalSearchTests.cs (hunk 2 above)
Comment (feordin) [reply test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/CanonicalSearchTests.cs:19]: Ah yes! sound good.

**Review by feordin — APPROVED** [review]

File: src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/UriSearchValue.cs (hunk 1 above)
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Core/Features/Search/SearchValues/UriSearchValue.cs]: Updated

**Review by johnstairs — APPROVED** [review]

**Pull request merged**